File: Ruby/lib/civitas-ruby/casilla.rb
require_relative 'tipo_sorpresa'
require_relative 'tipo_casilla'
require_relative 'titulo_propiedad'
require_relative 'mazo_sorpresas'

module Civitas
  class Casilla
    @@carcel

    def initialize(cantidad, num_casilla_carcel, mazo, nombre, titulo, tipo)
      @importe = cantidad
      @nombre = nombre
      @tipo = tipo
      @titulo_propiedad = titulo
      @mazo = mazo
      @@carcel = num_casilla_carcel if num_casilla_carcel != nil
    end

    attr_reader :nombre, :titulo_propiedad

    def self.new_descanso(nombre)
      new(nil, nil, nil, nombre, nil, TipoCasilla::DESCANSO)
    end

    def self.new_calle(titulo)
      new(nil, nil, nil, titulo.nombre, titulo, TipoCasilla::CALLE)
    end

    def self.new_impuesto(cantidad, nombre)
      new(cantidad, nil, nil, nombre, nil, TipoCasilla::IMPUESTO)
    end

    def self.new_juez(num_casilla_carcel, nombre)
      new(nil, num_casilla_carcel, nil, nombre, nil, TipoCasilla::JUEZ)
    end

    def self.new_sorpresa(mazo, nombre)
      new(nil, nil, mazo, nombre, nil, TipoCasilla::SORPRESA)
    end

    private_class_method :new

    def jugador_correcto(iactual, todos)
      iactual < todos.size
    end

    def recibe_jugador(iactual, todos)
      case @tipo
      when TipoCasilla::CALLE
        recibe_jugador_calle(iactual, todos)
      when TipoCasilla::IMPUESTO
        recibe_jugador_impuesto(iactual, todos)
      when TipoCasilla::JUEZ
        recibe_jugador_juez(iactual, todos)
      when TipoCasilla::SORPRESA
        recibe_jugador_sorpresa(iactual, todos)
      else
        informe(iactual, todos)
      end
    end

    def to_s
      "Nombre: #{@nombre}
      Tipo: #{@tipo}"
    end

    private

    def informe(iactual, todos)
      Diario.instance.ocurre_evento("#{todos[iactual].nombre} ha caído en 
      #{to_s}")
    end

    def recibe_jugador_calle(iactual, todos)
      if jugador_correcto(iactual, todos)
        informe(iactual, todos)
        jugador = todos[iactual]
        if !@titulo_propiedad.tiene_propietario
          jugador.puede_comprar_casilla
        else
          puts "Hola deberías tramitar el alquiler hermanazo"
          @titulo_propiedad.tramitar_alquiler(jugador)
        end
      end
    end

    def recibe_jugador_impuesto(iactual, todos)
      if jugador_correcto(iactual, todos)
        informe(iactual, todos)
        todos[iactual].paga_impuesto(@importe)
      end
    end

    def recibe_jugador_juez(iactual, todos)
      if jugador_correcto(iactual, todos)
        informe(iactual, todos)
        todos[iactual].encarcelar(@@carcel)
      end
    end

    def recibe_jugador_sorpresa(iactual, todos)
      if jugador_correcto(iactual, todos)
        sorpresa = @mazo.siguiente
        informe(iactual, todos)
        sorpresa.aplicar_a_jugador(iactual, todos)
      end
    end

  end
end

File: Java/src/civitas/Casilla.java
package civitas;

import java.util.ArrayList;

public class Casilla {
    
  private String nombre ;
  private static int carcel;
  private float importe;
  private TipoCasilla tipo;
  private TituloPropiedad tituloPropiedad;
  private MazoSorpresas mazoSorpresas;

  Casilla(String n) {
    init();
    nombre = n;
    tipo = TipoCasilla.DESCANSO;
  }

  Casilla(TituloPropiedad titulo) {
    init();
    tituloPropiedad = titulo;
    nombre = titulo.getNombre();
    tipo = TipoCasilla.CALLE;
  }

  Casilla(float cantidad, String n) {
    init();
    importe = cantidad;
    nombre = n;
    tipo = TipoCasilla.IMPUESTO;
  }

  Casilla(int numCasillaCarcel, String n) {
    init();
    carcel = numCasillaCarcel;
    nombre = n; 
    tipo = TipoCasilla.JUEZ;
  }

  Casilla(MazoSorpresas mazo, String n) {
    init();  
    mazoSorpresas = mazo;
    nombre = n;
    tipo = TipoCasilla.SORPRESA;
  }

  public String getNombre() {
    return nombre;
  }

  TituloPropiedad getTituloPropiedad() {
    return tituloPropiedad;
  }
  
  private void informe(int iactual, ArrayList<Jugador> todos) {
    Diario.getInstance().ocurreEvento("El jugador " + todos.get(iactual).getNombre() + " ha caído en la casilla " + getNombre());
  }

  private void init() {
    nombre = "";
    importe = 0;
    tipo = null;
    tituloPropiedad = null;
    mazoSorpresas = null;
  }

  public Boolean jugadorCorrecto(int iactual, ArrayList<Jugador> todos) {
    return(0 <= iactual && iactual<todos.size());
  }

  void recibeJugador(int iactual, ArrayList<Jugador> todos) {
    switch (tipo) {
      case CALLE:
        recibeJugador_calle(iactual, todos);
        break;
      case IMPUESTO:
        recibeJugador_impuesto(iactual, todos);
        break;
      case JUEZ:
        recibeJugador_juez(iactual, todos);
        break;
      case SORPRESA:
        recibeJugador_sorpresa(iactual, todos);
        break;
      default:
        informe(iactual, todos);
        break;
    }
  }

  private void recibeJugador_calle(int iactual, ArrayList<Jugador> todos) {
    if (jugadorCorrecto(iactual, todos)) {
      informe(iactual, todos);
      Jugador jugador = todos.get(iactual);
      if (!tituloPropiedad.tienePropietario()) {
        jugador.puedeComprarCasilla();
      }
      else {
        tituloPropiedad.tramitarAlquiler(jugador);
      }
    }  
  }

  private void recibeJugador_impuesto(int iactual, ArrayList<Jugador> todos) {
    if (jugadorCorrecto(iactual, todos)) {
      informe(iactual, todos);
      todos.get(iactual).pagaImpuesto(importe);
    }
  }

  private void recibeJugador_juez(int iactual, ArrayList<Jugador> todos) {
    if (jugadorCorrecto(iactual, todos)) {
      informe(iactual, todos);
      //System.err.println("Estoy en casilla.recibe_juez. Carcel en "+carcel);
      todos.get(iactual).encarcelar(carcel);
    }
  }

  private void recibeJugador_sorpresa(int iactual, ArrayList<Jugador> todos) {
    if (jugadorCorrecto(iactual, todos)) {
      Sorpresa sorpresa = mazoSorpresas.siguiente();
      informe(iactual, todos);
      sorpresa.aplicarAJugador(iactual, todos);
    }
  }

  public String toString() {
    String s = 
      "\n  Nombre: " + nombre +
      "\n  Tipo de la casilla: "+ tipo; 
    return s;
  }
}
